File: apps/mobile/modules/hither-menu/ios/HitherMenuModule.swift
import ExpoModulesCore
import UIKit

final class HitherMenuView: ExpoView {
  let button = UIButton(type: .custom)
  let onSelect = EventDispatcher()
  var items: [[String: String]] = [] {
    didSet { rebuildMenu() }
  }

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = false
    button.backgroundColor = .clear
    button.showsMenuAsPrimaryAction = true
    button.changesSelectionAsPrimaryAction = false
    addSubview(button)
    button.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      button.leadingAnchor.constraint(equalTo: leadingAnchor),
      button.trailingAnchor.constraint(equalTo: trailingAnchor),
      button.topAnchor.constraint(equalTo: topAnchor),
      button.bottomAnchor.constraint(equalTo: bottomAnchor),
    ])
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    bringSubviewToFront(button)
  }

  private func rebuildMenu() {
    let actions = items.compactMap { item -> UIAction? in
      guard let id = item["id"], let title = item["title"] else { return nil }
      let subtitle = item["subtitle"]
      let selected = item["selected"] == "true"
      let action = UIAction(
        title: title,
        subtitle: subtitle,
        identifier: UIAction.Identifier(id),
        state: selected ? .on : .off
      ) { [weak self] _ in
        self?.onSelect(["id": id])
      }
      return action
    }
    button.menu = UIMenu(title: "", options: [], children: actions)
  }
}

public class HitherMenuModule: Module {
  public func definition() -> ModuleDefinition {
    Name("HitherMenu")

    View(HitherMenuView.self) {
      Events("onSelect")
      Prop("items") { (view: HitherMenuView, items: [[String: String]]) in
        view.items = items
      }
    }
  }
}

File: apps/mobile/modules/hither-menu/android/src/main/java/expo/modules/hithermenu/HitherMenuModule.kt
package expo.modules.hithermenu

import android.content.Context
import android.view.Gravity
import android.widget.PopupMenu
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class HitherMenuView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  var items: List<Map<String, String>> = emptyList()
  private val onSelect by EventDispatcher()

  init {
    isClickable = true
    setOnClickListener {
      val popup = PopupMenu(context, this, Gravity.START)
      items.forEachIndexed { index, item ->
        popup.menu.add(0, index, index, item["title"] ?: "")
      }
      popup.setOnMenuItemClickListener { menuItem ->
        val id = items.getOrNull(menuItem.itemId)?.get("id") ?: return@setOnMenuItemClickListener false
        onSelect(mapOf("id" to id))
        true
      }
      popup.show()
    }
  }
}

class HitherMenuModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("HitherMenu")

    View(HitherMenuView::class) {
      Events("onSelect")
      Prop("items") { view: HitherMenuView, items: List<Map<String, String>> ->
        view.items = items
      }
    }
  }
}
